clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -592,9 +602,12 @@ private void resetOffsetIfNeeded(TopicPartition partition, Long requestedResetTi
         } else if (!requestedResetTimestamp.equals(offsetResetStrategyTimestamp(partition))) {
             log.debug("Skipping reset of partition {} since an alternative reset has been requested", partition);
         } else {
-            log.info("Resetting offset for partition {} to offset {}.", partition, offsetData.offset);
+            ConsumerMetadata.LeaderAndEpoch leaderAndEpoch = metadata.leaderAndEpoch(partition);
```

Comment:
Ok, so for the ListOffset response here, we should only update the position if the returned epoch matches that of the Metadata? And if not, just retry? (via `SubscriptionState#resetFailed`?)

Reply:
This gets a bit confusing. I think we somehow have to remember which epoch was sent in the ListOffset request. What I think we should be checking is whether the epoch in `FetchPosition.currentLeader` matches what was sent in the request. This would ensure that there was not a leader change while the ListOffset request was inflight.